### ISSUE TITLE
How to join topics onto a dataframe based on a unique document id

### ISSUE THREAD

**Issue by leifericf** [MaartenGr/BERTopic#564]
I have Pandas dataframe with many columns:

```
                             count  unique
starRating                286817.0     NaN
createTime                  286817  286816
updateTime                  286817  286816
reviewer.profilePhotoUrl    286817  235408
reviewer.displayName        286817  211631
reviewReply.comment            292     196
reviewReply.updateTime         292     292
comment_translated           91604   72956
comment_original             91604   76659
account_id                  286817       4
location_id                 286817     428
review_id                   286817  286756
account_name                286817       4
location_code               286817     428
location_name               286817     168
```

Each row represents a review with a unique id (`review_id`).

The column `comment_original` contains the document I will feed to BERTopic:

```python
df = df[df["comment_original"].notnull()]
docs = df["comment_original"].tolist()
topic_model = BERTopic(language="multilingual", verbose=True)
topic_model.fit_transform(docs)
topic_model.save("models/multilingual")
```

The other columns are dimensions I will use for my data visualization app created with Dash.

After creating a model with BERTopic, I want to assign the topics back to a new column in the original dataframe so that I can expose them in my Dash app in the proper context. I want to implement functionality in my Dash app so the user can drill down into a selected topic to view the underlying terms and documents when necessary.

Is it possible to pass the `review_id` into BERTopic, so that I can join the topics back onto the original dataframe?

I imagine the result as one dataframe with all data required for the Dash app, including BERTopic output.

**Comment by MaartenGr** [comment]
> After creating a model with BERTopic, I want to assign the topics back to a new column in the original dataframe so that I can expose them in my Dash app in the proper context. 

The most straightforward approach of doing this is as follows:

```python
# Train and save the model
df = df[df["comment_original"].notnull()]
docs = df["comment_original"].tolist()
topic_model = BERTopic(language="multilingual", verbose=True)
probs, topics = topic_model.fit_transform(docs)
topic_model.save("models/multilingual")

# Assign topics to the original dataframe
df["Topic"] = topics
```

This should work as each topic in `topics` corresponds to a document in `docs`.

**Comment by leifericf** [comment]
Thanks again for the help, @MaartenGr!

Does that rely on the documents being unique, or does BERTopic ensure that the row indexes of the topics it creates are the same as the document indexes passed into it? I.e., there will always be the same number of topics and documents.

Since these are textual reviews written by customers, they are likely to be unique, but there is no guarantee. There are many single-word reviews like "good," so I thought there might not be a 1:1 match between topics and documents.

**Comment by leifericf** [comment]
I believe I got it working like this:

```python
import pandas as pd
import data_prep
from bertopic import BERTopic

TRAIN_MODEL = True
VISUALIZE_MODEL = False
DEBUG = True

if TRAIN_MODEL == True:
    # Get dataframe with all ratings and reviews (including those without comments)
    df = data_prep.get_reviews_dataframe()

    # Filter out reviews which do not include comments
    comments = df[df["comment_original"].notnull()]

    # Train and save model
    topic_model = BERTopic(language="multilingual", verbose=True)
    topics, probs = topic_model.fit_transform(comments["comment_original"].tolist())
    topic_model.save("models/multilingual")

    # Append topic ids and probabilities to dataframe
    comments["topic_id"] = topics
    comments["topic_probability"] = probs

    # Merge topic ids and probabilities into big dataframe containing all reviews
    df = pd.merge(
        df, comments, how="left", on="review_id", suffixes=("", "_y"), validate="1:1"
    )

    # Drop duplicate columns from dataframe
    df.drop(df.filter(regex="_y$").columns.tolist(), axis=1, inplace=True)

    # Prints to verify that comments have been successfully merged into original dataframe
    if DEBUG == True:
        df_comments_with_topics = df[df["comment_original"].notnull()]
        print(
            df_comments_with_topics[
                ["review_id", "comment_original", "topic_id", "topic_probability"]
            ].sample(10)
        )
```

The print statement shows that the topics and probabilities are indeed merged into the main dataframe:

```
                                                review_id                                   comment_original  topic_id  topic_probability
144889  AbFvOqk5CpFEMiTiBi4DSRyojwGuhvv9cAtcB4r7e5i-Nq...   Åpent og oversiktlig butikk. Butikk ansatte e...      -1.0           0.000000
39744   AbFvOqkrAzKXdxzvjkaHegNCrfk8EZrQPdQ6Nx2Fc0gMbD...   Jag fick bra hjälp och köpte dubbel så mycket...      -1.0           0.000000
272361  AbFvOqkt9MNoOX3ditwWG-qGx3C6thUqzJaar-RT1EnF_9...                  Mukavan pieni. Ei suuria jonoja👌🏼      -1.0           0.000000
117798  AbFvOqkoCUZWh_dRlrjg0XasUNuwvCJcUfiNGkbzcWd6-N...                                         Väldig bra     782.0           1.000000
206393  AbFvOqk7TGS_-Yfg7w-hd37WQC9AMj_QlLxeLIy0imODdO...   Bedrövliga! Köpte precis en dammsugare, återl...      93.0           0.628041
141872  AbFvOqknCW1JiODtlZMkjlIqmG2oerwD3uN0B-i3Bql_8v...   Jag bara kollabde, gav ett random nummer stjä...     329.0           0.477420
96405   AbFvOqlNR-C02aW1cwUa-eI0QlEcQJ5OvUkncykSoi0q_H...                                 Kjempeflott plass!      77.0           0.699138
1209    AbFvOqmLN_b8XWl1XluYa2_3lgP01v5THbuQFDHtrRUbVh...                 Hurtig god betjening rar personale      -1.0           0.000000
115687  AbFvOqmanwqTm8f_rsLK_UZWo22qF40ys7TYwMpD0gwDjZ...                                        Bra grejer!      -1.0           0.000000
269816  AbFvOqlooQizRoA7UlOk_oK0MTZK0afLgA4O2RuLHxA5Gi...                  Löytyvät tarvittavat kodinkoneet.      34.0           1.000000
```

However, I'm unsure whether the `topic_id` and `topic_probabilitiy` got associated with the correct `review_id`. I suppose that depends on whether the row indexes were aligned correctly for this part:

```python
# Append topic ids and probabilities to dataframe
comments["topic_id"] = topics
comments["topic_probability"] = probs
```

I'm unsure how to verify that, but I'm working on it now.

**Comment by MaartenGr** [comment]
Yes, this seems to be correct since the topic in `topics[0]` relates to the document in `comments["comment_original"].tolist()[0]`, the topic in `topics[1]`  to `comments["comment_original"].tolist()[1]`, etc. This means that the order from the input documents are kept and that you can assign them like so:

```python
# Append topic ids and probabilities to dataframe
comments["topic_id"] = topics
comments["topic_probability"] = probs
```

To check whether you did this correctly, you can use `topic_model.get_representative_docs()` to check whether the documents match the topics. Also, you can do a visual inspection of each topic to see if they makes sense with the documents that you assigned it to in the original dataframe.

**Comment by leifericf** [comment]
Excellent, thanks for confirming! It seems to be correct. I was afraid some concurrent computation under the hood might shuffle the document list during model training, but that doesn't seem to be the case.

**Comment by MaartenGr** [comment]
@RamyTheEngineer The documents should not get scattered around during the `.partial_fit`, so the topics that it outputs should be in the same order as the input documents.

**Comment by RamyTheEngineer** [comment]
Thank you for the clarification. So for example:
`topic_model.partial_fit(docs_1)
topic_model.partial_fit(docs_2)`

the output should display docs_1 rows followed by docs_2 rows?

One more question, I am not able to perform:
`topics ,probs = topic_model.partial_fit(docs_train)`
this code owuld give me this error: cannot unpack non-iterable BERTopic object
Is there any other way to display these multiple docs according to their original representation so I can map each row to its own topic and previously specified ID?

**Comment by RamyTheEngineer** [comment]
@MaartenGr Okay thank you so much 😄
